matchzoo/engine/base_transformer.py

```diff
@@ -0,0 +1,296 @@
+"""Contains the base Transformer class for  all transformers."""
+from __future__ import print_function
+from __future__ import absolute_import
+from __future__ import division
+
+import os
+import abc
+import typing
+import pickle
+import re
+import six
+from collections import Mapping
+
+from matchzoo import engine
+
+
+class BaseTransformer(abc.ABC):
+    """Base data transformer for MatchZoo.
+
+    # Methods:
+    ==========
+    fit_transform(X, y=None, **fit_params)
+
+    fit(X, y=None)
+
+    transform(X, copy=True)
+
+    """
+
+    _white_spaces = re.compile(r"\s\s+")
+
+    def __init__(self, params: typing.Optional[engine.TransformerParams]
+                 = None):
+        """:class:`BaseTransformer` constructor."""
+        self._params = params or self.get_default_params()
+
+    @classmethod
+    def get_default_params(cls) -> engine.TransformerParams:
+        """Return the default parameters for transformer."""
+        return engine.TransformerParams()
+
+    def guess_and_fill_missing_params(self):
+        """
+        Guess and fill missing parameters in :attribute:`params`.
+
+        Note: likely to be moved to a higher level API in the future.
+        """
+        if self._params['name'] is None:
+            self._params['name'] = self.__class__.__name__
+
+        if self._params['transformer_class'] is None:
+            self._params['transformer_class'] = self.__class__
+
+    @property
+    def params(self):
+        """:return: model parameters."""
+        return self._params
+
+    def _word_ngrams(self, tokens, stop_words=None):
+        """Turn tokens into a sequence of n-grams."""
+        # handle stop words
+        if stop_words is not None:
+            tokens = [w for w in tokens if w not in stop_words]
+
+        # handle token n-grams
+        min_n, max_n = self._params['ngram_range']
+        if max_n != 1:
+            original_tokens = tokens
+            if min_n == 1:
+                # no need to do any slicing for unigrams
+                # just iterate through the original tokens
+                tokens = list(original_tokens)
+                min_n += 1
+            else:
+                tokens = []
+
+            n_original_tokens = len(original_tokens)
+
+            # bind method outside of loop to reduce overhead
+            tokens_append = tokens.append
+            space_join = " ".join
+
+            for n in range(min_n, min(max_n + 1, n_original_tokens + 1)):
+                for i in range(n_original_tokens - n + 1):
+                    tokens_append(space_join(original_tokens[i: i + n]))
```

Comment:
I do not prefer this type of `+1`, `-1` `original_tokens[i:i+n]` not pythonic and easy to trigger IndexOutOfRangeError.